packages/eslint-plugin-turbo/lib/utils/calculate-inputs.ts

```diff
@@ -0,0 +1,449 @@
+    this.allConfigs = getWorkspaceConfigs(cwd);
+    this.projectRoot = this.allConfigs.find(
+      (workspaceConfig) => workspaceConfig.isWorkspaceRoot
+    );
+    this.projectWorkspaces = this.allConfigs.filter(
+      (workspaceConfig) => !workspaceConfig.isWorkspaceRoot
+    );
+
+    this._key = this.generateKey();
+    this._test = this.generateTestConfig();
+  }
+
+  valid(): boolean {
+    return this.allConfigs.length > 0;
+  }
+
+  generateKey(): ProjectKey {
+    let global: EnvironmentConfig = {
+      legacyConfig: [],
+      env: [],
+      passThroughEnv: null,
+      dotEnv: null,
+    };
+    let globalTasks: {
+      [script: string]: EnvironmentConfig;
+    } = {};
+    let workspaceTasks: {
+      [workspace: string]: {
+        [script: string]: EnvironmentConfig;
+      };
+    } = {};
+
+    if (
+      this.projectRoot &&
+      this.projectRoot.turboConfig &&
+      !("extends" in this.projectRoot)
+    ) {
+      const rootTurboJson = this.projectRoot;
+
+      global = processGlobal(
+        this.projectRoot.workspacePath,
+        this.projectRoot.turboConfig
+      );
+
+      Object.entries(this.projectRoot.turboConfig.pipeline).forEach(
+        ([taskName, taskDefinition]) => {
+          const { workspaceName, scriptName } = getTaskAddress(taskName);
+          if (workspaceName) {
+            workspaceTasks[workspaceName] = workspaceTasks[workspaceName] || {};
+            workspaceTasks[workspaceName][scriptName] = processTask(
+              rootTurboJson.workspacePath,
+              taskDefinition
+            );
+          } else {
+            globalTasks[scriptName] = processTask(
+              rootTurboJson.workspacePath,
+              taskDefinition
+            );
+          }
+        }
+      );
+    }
+
+    this.projectWorkspaces.forEach((projectWorkspace) => {
+      if (!projectWorkspace.turboConfig) {
+        return;
+      }
+
+      Object.entries(projectWorkspace.turboConfig.pipeline).forEach(
+        ([taskName, taskDefinition]) => {
+          const { workspaceName: erroneousWorkspaceName, scriptName } =
+            getTaskAddress(taskName);
+          if (erroneousWorkspaceName) {
+            throw new Error(
+              "May not specify workspace name in non-root turbo.json"
+            );
+          }
+
+          const workspaceName = projectWorkspace.workspaceName;
+          workspaceTasks[workspaceName] = workspaceTasks[workspaceName] || {};
+          workspaceTasks[workspaceName][scriptName] = processTask(
+            projectWorkspace.workspacePath,
+            taskDefinition
+          );
+        }
+      );
+    });
+
+    return {
+      global,
+      globalTasks,
+      workspaceTasks,
+    };
+  }
+
+  getWorkspacePath(workspaceName: string): string | undefined {
+    return this.projectWorkspaces.find(
+      (workspaceConfig) => workspaceConfig.workspaceName === workspaceName
+    )?.workspacePath;
+  }
+
+  generateTestConfig(): ProjectTests {
```

Comment:
Look away from this monstrosity lest Zalgo consume your soul. I got in a fight with TypeScript and lost.